Rosalina Teleport downb

chrispo-git opened the pull request:
Rosalina downb now acts as a teleport, swapping her and Luma's position.
Frames 17-24 - invulnerable (teleporting)
Frame 38 - Actionable

Rosalina cannot use downb for 300 frames (5 seconds) afterwards, and her being able to use it again is indicated by a flash on her hands and on the HUD. If you're too far away, Rosalina cannot teleport.

https://user-images.githubusercontent.com/69305550/177057739-f0bee2e4-5758-48f0-8e09-984b94f8e265.mp4

chrispo-git committed:
Rosalina Teleport

Rosalina downb now acts as a teleport, swapping her and Luma's position.
Frames 17-24 - invulnerable (teleporting)
Frame 38 - Actionable

Rosalina cannot use downb for 300 frames (5 seconds) afterwards, and her being able to use it again is indicated by a flash on her hands and on the HUD

WuBoytH commented on fighters/rosetta/src/opff.rs:
```diff
@@ -26,6 +83,51 @@ pub fn rosetta_frame_wrapper(fighter: &mut smash::lua2cpp::L2CFighterCommon) {
+        moveset(fighter, &mut *info.boma, info.id, info.cat, info.status_kind, info.situation_kind, info.motion_kind.hash, info.stick_x, info.stick_y, info.facing, info.frame);
+    }
+}
+#[smashline::weapon_frame_callback]
```
Weapons can have their own specific fighter frames, you can change the attribute to this `#[weapon_frame( agent = WEAPON_KIND_ROSETTA_TICO )]`

WuBoytH reviewed (changes requested):
Look at my comment about Luma's OPFF. Other than that, it should need playtesting (and a second opinion from someone else) for further adjustments before approval.

chrispo-git committed:
Teleport fixes and rebalancing

Luma is now in his own opwf.
As requested in the discord, there is now a max distance you can initiate teleport from (130 units horizontally, 70 units vertically) to prevent it being used as a powerful offstage recovery tool and landing option.

chrispo-git committed:
Invisibility Fix

Teleporting right on luma death no longer causes rosalina to remain invisible, and maintains her actionability of 38

jobrien97 commented:
Please change the indicator when teleport is recharged to our common gimmick flash effect found in common/src/opff/gimmick.rs

chrispo-git committed:
Gimmick Flash

Now correctly uses gimmick flash when recharge time is done

WuBoytH commented on fighters/rosetta/src/opff.rs:
```diff
@@ -12,8 +12,80 @@ unsafe fn launch_star_cancel(boma: &mut BattleObjectModuleAccessor, status_kind:
+					};
+				};
+				if VarModule::get_int(fighter.battle_object, vars::rosetta::instance::COOLDOWN) > 0 {
+					VarModule::set_int(fighter.battle_object, vars::rosetta::instance::COOLDOWN,  VarModule::get_int(fighter.battle_object, vars::rosetta::instance::COOLDOWN)-1);
```
you can just do `VarModule::dec_int(fighter.battle_object, vars::rosetta::instance::COOLDOWN);` to decrement this var by 1.

WuBoytH commented on fighters/rosetta/src/opff.rs:
```diff
@@ -12,8 +12,80 @@ unsafe fn launch_star_cancel(boma: &mut BattleObjectModuleAccessor, status_kind:
+					VarModule::set_int(fighter.battle_object, vars::rosetta::instance::COOLDOWN, 0);
+					VarModule::off_flag(boma.object(), vars::rosetta::instance::IS_TICO_DEAD);
+				};
+				VarModule::set_int(fighter.battle_object, vars::rosetta::instance::TICO_RAYCAST, 
```
I would appreciate that you don't nest VarModule calls for readability purposes.
`let rosa_x = ...;`, `let tico_x = ...;`, etc. Please check this for the rest of the function as well.

WuBoytH reviewed (changes requested):
Just a few smaller, organizational changes are needed before we can merge this in. Once it *does* get merged in, I may try and reimplement this as a status script instead.

chrispo-git committed:
Organizational changes

Now uses dec_int for the cooldown
No longer nests varmodule calls

WuBoytH reviewed (approved):
LGTM
